src/refract/JsonSchema.cc

```diff
@@ -0,0 +1,290 @@
+//
+//  refract/JsonSchema.cc
+//  librefract
+//
+//  Created by Thomas Jandecka on 16/02/2018
+//  Copyright (c) 2018 Apiary Inc. All rights reserved.
+//
+
+#include "JsonSchema.h"
+
+#include <cassert>
+#include <bitset>
+#include "Element.h"
+#include "../ElementData.h"
+#include "Utils.h"
+#include "../utils/log/Trivial.h"
+
+#include "VisitorUtils.h" // GetValue
+
+using namespace refract;
+using namespace schema;
+using namespace drafter::utils;
+using namespace drafter::utils::log;
+
+namespace
+{ // API Elements tools
+    so::String instanciate(const StringElement& e)
+    {
+        assert(!e.empty());
+        return so::String{ e.get().get() };
+    }
+
+    so::Number instanciate(const NumberElement& e)
+    {
+        assert(!e.empty());
+        return so::Number{ e.get().get() };
+    }
+
+    so::Value instanciate(const BooleanElement& e)
+    {
+        assert(!e.empty());
+        return e.get() ? //
+            so::Value{ in_place_type<so::True>{} } :
+            so::Value{ in_place_type<so::False>{} };
+    }
+
+    bool has_fixed_attr(const IElement& e)
+    {
+        return e.attributes().end() != e.attributes().find("fixed");
+    }
+
+    bool has_fixed_type_attr(const IElement& e)
+    {
+        return e.attributes().end() != e.attributes().find("fixedType");
+    }
+
+    template <typename T>
+    const T* get_if_default_attr(const T& e)
+    {
+        auto it = e.attributes().find("default");
+        if (e.attributes().end() != it)
+            return TypeQueryVisitor::as<const T>(it->second.get());
+        return nullptr;
+    }
+
+    const dsd::Array& get_enumerations(const EnumElement& e)
+    {
+        auto resultIt = e.attributes().find("enumerations");
+        assert(e.attributes().end() != resultIt);
+
+        const auto enums = TypeQueryVisitor::as<const ArrayElement>(resultIt->second.get());
+        assert(enums);
+        assert(!enums->empty());
+
+        return enums->get();
+    }
+}
+
+namespace
+{
+    using flags = std::bitset<1>;
+    constexpr std::size_t fixed_flag = 0;
+}
+
+namespace
+{ // JSON Schema tools
+    so::Object& add_schema_version(so::Object& schema)
+    {
+        schema.data.emplace_back("$schema", so::String{ "http://json-schema.org/draft-04/schema#" });
```

Comment:
💭 We should start thinking about JSON Schema draft 7 😬. Since we don't delve into validations and other features of JSON Schema, perhaps our generated schemas would already be compatible.

Reply:
Agreed. We could do that after `JsonSchema.cc` is integrated and changes in fixtures are reviewed, to keep diffs as sane as possible.
:thought_balloon:  A commit in the future changing both the URI to `draft-06` and `"enum":[foo]` to `"const": foo`.